dali/operators/random/beta_distribution.h

```diff
@@ -0,0 +1,169 @@
+// Copyright (c) 2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_OPERATORS_RANDOM_BETA_DISTRIBUTION_H_
+#define DALI_OPERATORS_RANDOM_BETA_DISTRIBUTION_H_
+
+#include <random>
+#include <vector>
+#include "dali/operators/random/rng_base.h"
+#include "dali/pipeline/operator/arg_helper.h"
+
+#define DALI_BETA_DIST_TYPES float, double
+
+namespace dali {
+
+template <typename T>
+struct BetaDistributionImpl {
+  BetaDistributionImpl() : BetaDistributionImpl{1, 1} {}
+
+  explicit BetaDistributionImpl(T alpha, T beta)
+      : has_small_param_{alpha < 1 && beta < 1},
+        alpha_{alpha},
+        beta_{beta},
+        b_div_a_{beta_ / alpha_},
+        a_{has_small_param_ ? alpha_ + 1 : alpha_},
+        b_{has_small_param_ ? beta_ + 1 : beta_},
+        exp_{} {}
+
+  template <typename Generator>
+  T Generate(Generator &st) {
+    return has_small_param_ ? GenerateGammaExp(st) : GenerateGamma(st);
+  }
+
+ private:
+  template <typename Generator>
+  T GenerateGamma(Generator &st) {
+    // https://en.wikipedia.org/wiki/Beta_distribution#Random_variate_generation
+    T a = a_(st);
+    T b = b_(st);
+    return a / (a + b);
+  }
+
+  template <typename Generator>
+  T GenerateGammaExp(Generator &st) {
+    assert(alpha_ < 1 && beta_ < 1);
+    // For alpha >= 1 and x in [0, 1], Gamma(alpha).cdf(x) <= Exp(1).cdf(x),
+    // i.e., the probability of sampling [0, x] is less than 1 - exp(-x) <= x,
+    // so the probability of the denominator in `A / (A + B)` being rounded to zero
+    // is negligable.
+    // However, for alpha < 1, the gamma distribution has high density near 0.
+    // Here, we use the fact that for X ~ Gamma(alpha + 1), U ~ Uniform(0, 1)^(1/alpha),
+    // the `A = X * U` has Gamma(alpha) distribution (see Luc Devroye, Non-Uniform Random
+    // Variate Generation, p. 182)
+    // We can compute A / (A + B) = X * U_a / (X * U_a + Y * U_b) =
+    // (X * U_a / max(U_a, U_b)) / (X * U_a / max(U_a, U_b) + Y * U_b / max(U_a, U_b)).
+    // This way, we have either X or Y in the denominator sampled from "safe" Gamma
+    // with params > 1.
+    // The Uniform(0, 1)^(1/alpha) is concentrated near zero, for this reason
+    // the U_a / max(U_a, U_b), U_b / max(U_a, U_b) are computed in logarithmic scale.
```

Comment:
For the future reference, even doubles are not satisfactory for direct usage with just Gamma nor non-log scale. Getting exactly zero for one of the summand jumps to around 48% for alpha=1e-3.

Below, mf is double's nextafter(1., 0.).
![image (25)](https://github.com/NVIDIA/DALI/assets/11878086/b7a0c990-a8f1-4446-a28a-34b0bbbf9b22)